source-pendo/source_pendo/api.py

```diff
@@ -0,0 +1,185 @@
+                        }
+                    }
+                },
+                # Only return events that have occurred since the last event we've seen.
+                {
+                    "filter": f"guideTimestamp > {last_ts}"
+                },
+                {
+                    "sort": ["guideTimestamp"]
+                },
+                {
+                    "limit": RESPONSE_LIMIT
+                }
+            ]
+        }
+    }
+
+    response = EventResponse.model_validate_json(await http.request(log, url, method="POST", json=body))
+    events = response.results
+
+    doc_count = 0
+    for event in events:
+        updated_at = getattr(event, "guideTimestamp", -1)
+        # Due to how we're querying the API with the "sort" and "filter" operators, 
+        # we don't expect to receive documents out of order.
+        if updated_at < last_ts:
+            raise RuntimeError(
+                f"Received events out of time order: Current event timestamp is {updated_at} vs. prior timestamp {last_ts}"
+            )
+
+        doc_count += 1
+        last_ts = updated_at
+
+        event.meta_ = model.Meta(op="c")
+        yield event
+
+    if doc_count == 0:
+        # If there were no documents, don't update the cursor.
+        return
+    
+    # If there was at least one document, update the cursor.
+    yield last_ts
+
+
+async def fetch_aggregated_events(
+        http: HTTPSession,
+        entity: str,
+        model: type[BaseDocument],
+        log: Logger,
+        log_cursor: LogCursor,
+) -> AsyncGenerator[PageEvent | FeatureEvent | TrackEvent | GuideEvent | PollEvent | LogCursor, None]:
+    assert isinstance(log_cursor, int)
+    url = f"{API}/aggregation"
+    last_ts = log_cursor
+    current_hour = pendulum.now("UTC").set(minute=0, second=0).int_timestamp * 1000
+
+    # Pendo's event aggregation endpoints require a JSON body with various fields.
+    # See https://engageapi.pendo.io/#f42c416f-bdd6-4082-b86c-1be3fe8401bf. We use
+    # their filtering and sorting capabilities to only return documents we haven't
+    # seen yet in ascending order. Each document returned represents an aggregation 
+    # of events in a specific hour.
+    body = {
+        "response": {
+            "mimeType": "application/json"
+        },
+        "request": {
+            "pipeline": [
+                {
+                    "source": {
+                        entity: None,
+                        "timeSeries": {
+                            "period": "hourRange",
+                            "first": f"{last_ts}",
+                            "last": "now()"
+                        }
+                    }
+                },
+                # Only return aggregates that have been updated since the last aggregate 
+                # we've seen AND aren't "in-progress" during the current hour.
+                {
+                    "filter": f"lastTime > {last_ts} && hour < {current_hour}"
```

Comment:
I don't think this is 100% bullet-proof, and it's because it involves comparisons between distributed clocks. In an extreme case, imagine the connector's system clock was 5 minutes fast. At 2:01 it could think it was fine to start getting results for the 1:00 hour, but the actual time is 1:56 and that would get an incomplete result for that hour.

Would it work to just always ignore the most recent aggregation result, and only output documents for the "N-1"'th aggregates? This would cause additional delay in outputting a document for the "latest" hour, but considering how this API seems to behave I don't think that would be a huge issue.

Reply:
I think I understand the issue. So in that extreme case, it's possible we emit aggregate documents more than once per hour for a specific aggregate.  Since these streams are for multiple different aggregate, IP address, and page/feature/track combos, not outputting the "N-1"th document won't address this issue. To solve this, we'd have to keep track of each unique aggregate combo we see in an hour & avoid emitting the N-1 doc for each of those combos. 

That's really tedious/difficult to implement. Since this is unlikely to be an issue, I just added a 15 minute delay when we check for the previous hour's aggregates - ex: at 2:15 we check for the 1:00 hour aggregates.